bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+                }
+            } else {
+                unregister(r.actions);
+                unregister(r.conditions);
+            }
+        } else {
+            unregister(r.conditions);
+        }
+    }
+
+    private <T extends Module> boolean register(List<T> modules, boolean all) {
+        boolean result = false;
+        if (modules != null) {
+            for (Iterator<T> it = modules.iterator(); it.hasNext();) {
+                T t = it.next();
+                try {
+                    ModuleHandler moduleHandler = getModuleHandler(t);
+                    ((ModuleImpl) t).setModuleHandler(moduleHandler);
+                    result = true;
+                } catch (Throwable e) {
+                    log.warn(
+                            "Missing handler: " + t.getTypeUID() + ", for the module: " + t.getId() + ". "
+                                    + e.getMessage(), e);
+                    if (all) {
+                        return false;
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    private <T extends Module> void unregister(List<T> modules) {
+        if (modules != null) {
+            for (Iterator<T> it = modules.iterator(); it.hasNext();) {
+                ModuleImpl m = (ModuleImpl) it.next();
+                ModuleHandler moduleHandler = m.getModuleHandler();
+                if (moduleHandler != null) {
+                    moduleHandler.dispose();
+                    m.setModuleHandler(null);
+                }
+            }
+        }
+    }
+
+    private void startRule(RuleImpl r) {
+        RuleEngineCallback reCallback = RuleEngineCallbackImpl.getInstance(r, this);
+        for (Iterator<Trigger> it = r.triggers.iterator(); it.hasNext();) {
+            TriggerImpl t = (TriggerImpl) it.next();
+            TriggerHandler triggerHandler = t.getModuleHandler();
+            triggerHandler.setRuleEngineCallback(reCallback);
+        }
+    }
+
+    /**
+     * @param id
+     */
+    private ModuleHandler getModuleHandler(Module m) {
+        String mtId = m.getTypeUID();
+        if (mtId == null) {
+            throw new IllegalArgumentException("Invalid module type id. It must not be null!");
+        }
+        int idx = mtId.indexOf(MODULE_TYPE_SEPARATOR);
+        if (idx != -1) {
+            mtId = mtId.substring(0, idx);
+        }
+        ModuleHandlerFactory msf = moduleHandlerFactories.get(mtId);
+        if (msf == null) {
+            throw new IllegalArgumentException("Invalid module handler factory: " + mtId);
+        }
+        return msf.create(m);
+    }
+
+    protected RuleImpl removeRule(String id) {
+        return rules.remove(id);
+    }
+
+    protected RuleImpl getRule(String rUID) {
+        return rules.get(rUID);
+    }
+
+    protected void setRuleEnable(String rUID, boolean isEnabled) {
+        RuleImpl r = rules.get(rUID);
+        if (r != null) {
+            r.setEnabled(isEnabled);
```

Comment:
1) removeRule() is changed to:
    public synchronized boolean removeRule(String ruleUID) {
        RuleImpl r = re.removeRule(ruleUID);
        if (r != null) {
            removeRule(r);
        }
        return false;
    }
where the removeRule(r) is a new added method method to the RuleRegistry. The removeRule(r) deletes rule from the storage. 

2) The method setRuleEnable() must call storeRule() to persist new rule state. Unfortunately it was not implemented yet.